internal/pkg/memorystore/table.go

```diff
@@ -85,18 +85,34 @@ func (t *Table) Add(key string, value interface{}) *Row {
 	return row
 }
 
+// Add a new item if it doesn't exist.
+func (t *Table) Upsert(key string, value interface{}) *Row {
+	t.Lock()
+	defer t.Unlock()
+
+	row := t.getInternal(key)
+	if row != nil {
+		t.deleteInternal(key)
+	}
```

Comment:
The row key will be updated or created by the t.rows map, extra check not needed.